Source/Swig/symbol.c

```diff
@@ -382,6 +382,26 @@ String *Swig_symbol_qualified_language_scopename(Symtab *n) {
   return result;
 }
 
+/* -----------------------------------------------------------------------------
+ * Swig_symbol_add_using()
+ * ----------------------------------------------------------------------------- */
+
+void Swig_symbol_add_using(String * name, String * uname, Node * n) {
+  Hash *h;
+  h = Swig_symbol_clookup(uname,0);
+  if (h && checkAttribute(h, "kind", "class")) {
+    String *qcurrent = Swig_symbol_qualifiedscopename(0);
+    if (qcurrent) {
+	Append(qcurrent, "::");
+	Append(qcurrent, name);
+    } else {
+	qcurrent = NewString(name);
+    }
+    Setattr(symtabs, qcurrent, n);
+    Delete(qcurrent);
+  }
+}
```

Comment:
I think this new function would benefit greatly from a documentation comment explaining what it does and why - I'm not entirely sure I understand what's going on here.  The description of `Swig_symbol_clookup()` is:

```
/* -----------------------------------------------------------------------------
 * Swig_symbol_clookup()
 *
 * Look up a symbol in the symbol table.   This uses the C name, not scripting
 * names.   Note: If we come across a using declaration, we follow it to
 * to get the real node. Any using directives are also followed (but this is
 * implemented in symbol_lookup()).
 * ----------------------------------------------------------------------------- */
```

From the code it seems we get the "real node" for the "using" symbol, and if that's a class we qualify a different name and add a symtab entry for that, but I'm unclear why we're doing that.  I'm also left unsure why this is only done for classes (in particular it seems kind can be `"struct"` which in C++ is just a class with default visibility of `public`, but why is this not done for other kinds too?)

Reply:
If I do `s/class/struct/` on `using_member.i` then the test fails, but all the classes in there only have public members, so that ought to make no difference:

```
$ make using_member.cpptest
if [ -f ./using_member_runme.py ]; then echo "checking python testcase using_member (with run test)" ; else echo "checking python testcase using_member" ; fi
checking python testcase using_member (with run test)
make -f ../../../Examples/Makefile SRCDIR='./' CXXSRCS='' SWIG_LIB_DIR='../../../Lib' SWIGEXE='../../../swig' INCLUDES='-I../../../Examples/test-suite' SWIGOPT='-outcurrentdir -I../../../Examples/test-suite' NOLINK=true TARGET='using_member' INTERFACEDIR='../' INTERFACE='using_member.i' python_cpp
make[1]: Entering directory '/home/olly/git/swig/Examples/test-suite/python'
env SWIG_LIB=../../../Lib  ../../../swig -python  -c++ -outcurrentdir -I../../../Examples/test-suite -o using_member_wrap.cxx ./../using_member.i
./../using_member.i:20: Warning 315: Nothing known about 'A::get'.
./../using_member.i:45: Warning 315: Nothing known about 'two::three::AA::great'.
g++ -c -fpic -I. -I./ -I/usr/include  using_member_wrap.cxx   -I../../../Examples/test-suite  -I/usr/include/python2.7 -I/usr/lib/python2.7/config
g++ -shared  -I/usr/include     using_member_wrap.o    -o _using_member.so
make[1]: Leaving directory '/home/olly/git/swig/Examples/test-suite/python'
pycodestyle --ignore=E252,E30,E402,E501,E731,W291,W391 using_member.py
if [ -f ./using_member_runme.py ]; then env LD_LIBRARY_PATH=.:$LD_LIBRARY_PATH PYTHONPATH=.:.:$PYTHONPATH  python using_member_runme.py; fi
Traceback (most recent call last):
  File "using_member_runme.py", line 4, in <module>
    assert b.get(int(1)) == 10
AssertionError
make: *** [Makefile:122: using_member.cpptest] Error 1
```

Reply:
The new nodes are specifically handling using declarations for class methods. Without this fix SWIG would simply not handle them and issue warnings like:
`Warning 315: Nothing known about 'A::get'`

Of course it should also handle struct methods, thanks for testing this out.

Do you think there is a better way to handle this?

Reply:
I added a brief description and support for structs.